### PR TITLE
Add support for multiple listen addresses

### PR DESCRIPTION
**Pull request by ShadowNinja**
- Ask the OS to assign an unused port instead of:
  - Using a static one (in tests, specifically `30001` and `30003`)
  - Trying a random one that might be used (client outgoing connections)
- **^** Fixes a few reported issues, but I only found #2344  with a cursory look.
- Make banning normalize IPv6-mapped IPv4 addresses (so you can't try `1.2.3.4` and then `::ffff:1.2.3.4`).
- Fix some horrible main menu code (multiple lookups, no popping, using `luaL_check*` for table fields).
- Move `socket.cpp` into `network` folder.
- Some style fixes.
- Lots of stuff that I don't remember right now.

To do:
- [ ] Maybe figure out a nicer format for the `--listen` argument (eg `--listen [::1]:0 --listen [::2]:0` instead of `--listen '[::1]:0 [::2]:0'`)

To do later:
- Fix how `gamedata.port` is stored as a string until just before it's used.

### CODE REVIEW

File: src/script/lua_api/l_server.cpp [hunk 1]
```diff
@@ -152,90 +148,71 @@ int ModApiServer::l_get_player_information(lua_State *L)
...
+	ERET(getServer(L)->getClientConInfo(player->peer_id, con::AVG_RTT,    &avg_rtt))
+	ERET(getServer(L)->getClientConInfo(player->peer_id, con::MIN_JITTER, &min_jitter))
+	ERET(getServer(L)->getClientConInfo(player->peer_id, con::MAX_JITTER, &max_jitter))
+	ERET(getServer(L)->getClientConInfo(player->peer_id, con::AVG_JITTER, &avg_jitter))
 
 	ERET(getServer(L)->getClientInfo(player->peer_id,
-										&state, &uptime, &ser_vers, &prot_vers,
-										&major, &minor, &patch, &vers_string))
+			&state, &uptime, &ser_vers, &prot_vers,
+			&major, &minor, &patch, &vers_string))
```
Comment (est31) [inline src/script/lua_api/l_server.cpp]: why two tabs?

Comment (ShadowNinja) [inline src/script/lua_api/l_server.cpp]: Because it looked nice to me that way.

**Comment by nerzhul** [comment]
@ShadowNinja for "Fix SQLite3 database statement initialization (because apparently that wasn't already done)." please push the fix with an atomic commit directly into master please and remove it from this PR

File: src/network/connection.cpp [hunk 2]
```diff
@@ -2102,17 +2109,16 @@ void ConnectionReceiveThread::receive()
 
 	unsigned int loop_count = 0;
 
-	/* first of all read packets from socket */
-	/* check for incoming data available */
-	while( (loop_count < 10) &&
-			(m_connection->m_udpSocket.WaitData(50))) {
+	// First read packets from socket
+	// Check for incoming data available
+	while ((loop_count < 10) && wait(100 * 1000, readset)) {
```
Comment (nerzhul) [inline src/network/connection.cpp]: maybe we could use a g_setting instead of a static 100*1000 (and cache it). And also 100 sec timeout is too large. Please set 30sec instead, like a classic TCP connection

**Comment by proller** [comment]
again small changes hides in huge space mess

File: src/network/connection.cpp (hunk 2 above)
Comment (ShadowNinja) [inline src/network/connection.cpp]: It's `100 * 1000` microseconds (us), that's `100` miliseconds (ms) or `0.1` seconds (s).
And it doesn't really make sense for this to be configurable.  You can't really optimize anything by changing it.

**Comment by ShadowNinja** [comment]
@nerzhul 3c6e112c289e73b8dfb8e1bcc74f9bd5c24b8f7d

**Comment by ShadowNinja** [comment]
@nerzhul http://jenkins.unix-experience.fr/job/minetest%20-%20FreeBSD%20-%20Official/474/ is apparently a jenkins or GitHub failure.  Please re-build t.

File: src/ban.cpp [hunk 3]
```diff
@@ -51,20 +49,18 @@ void BanManager::load()
...
-	if(is.good() == false)
-	{
+	if (!is) {
 		infostream<<"BanManager: failed loading from "<<m_banfilepath<<std::endl;
 		throw SerializationError("BanManager::load(): Couldn't open file");
 	}
 	
-	while(!is.eof() && is.good())
-	{
+	while (is) {
```
Comment (kwolekr) [inline src/ban.cpp]: Just because some standard classes have operator overloads doesn't mean you should use them!  Isn't this less expressive than the former condition?

File: src/game.cpp [hunk 4]
```diff
@@ -1912,44 +1916,21 @@ bool Game::initSound()
 	return true;
 }
 
-bool Game::createSingleplayerServer(const std::string map_dir,
-		const SubgameSpec &gamespec, u16 port, std::string *address)
+bool Game::createServer(const std::string map_dir,
+		const SubgameSpec &gamespec,
+		const std::string &listen)
```
Comment (kwolekr) [inline src/game.cpp]: Even if it's obvious that listen is an argument of the function in this context, we should still try to avoid shadowing symbols.

File: src/network/connection.cpp [hunk 5]
```diff
@@ -832,62 +831,30 @@ PeerHelper::PeerHelper() :
...
 	m_peer(peer)
 {
-	if (peer != NULL)
-	{
-		if (!peer->IncUseCount())
-		{
-			m_peer = 0;
-		}
-	}
+	if (!peer) return;
```
Comment (kwolekr) [inline src/network/connection.cpp]: Please don't add conditional bodies on the same line as the statement unless you're trying to hide something... this violates the Linux code style guidelines anyway (which ours inherit)

File: src/defaultsettings.cpp [hunk 6]
```diff
@@ -199,63 +203,59 @@ void set_default_settings(Settings *settings)
 	settings->setDefault("fallback_font_shadow", "1");
 	settings->setDefault("fallback_font_shadow_alpha", "128");
 
-	std::stringstream fontsize;
-	fontsize << TTF_DEFAULT_FONT_SIZE;
+	std::string font_size_str = to_string(TTF_DEFAULT_FONT_SIZE);
```
Comment (kwolekr) [inline src/defaultsettings.cpp]: How does this even compile?  to_string() is C++11
Shouldn't you just use ## to stringify a numerical constant?

File: src/network/socket.cpp [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+	}
+
+	// Do getaddrinfo()
+	int err = getaddrinfo(name.c_str(), NULL, &hints, &resolved);
+	if (err)
+		throw ResolveError(gai_strerror(err));
+
+	// Copy data
+	switch (resolved->ai_family) {
+	case AF_INET: {
```
Comment (kwolekr) [inline src/network/socket.cpp]: Shouldn't bodies inside of cases be like...
switch (foo) {
case FOO_BAR: {
    <sdfgh>;
    break;
}
case FOO_BAZ: {
 ...
?

File: src/network/connection.cpp [hunk 8]
```diff
@@ -2089,12 +1981,57 @@ void * ConnectionReceiveThread::Thread()
 	return NULL;
 }
 
+bool ConnectionReceiveThread::wait(int timeout_us, fd_set &readset) {
```
Comment (kwolekr) [inline src/network/connection.cpp]: Non-const reference parameter here

File: src/network/connection.cpp [hunk 9]
```diff
@@ -2089,12 +1981,57 @@ void * ConnectionReceiveThread::Thread()
...
+	// Initialize the set
+	FD_ZERO(&readset);
+	int max_fd = 0;
+	for (std::vector<UDPSocket *>::const_iterator
```
Comment (kwolekr) [inline src/network/connection.cpp]: Shouldn't this use array-iteration syntax? (for (int i = 0; i != m_sockets.size(); i++) { m_sockets[i]->whatever(); ...

File: src/defaultsettings.cpp (hunk 6 above)
Comment (ShadowNinja) [inline src/defaultsettings.cpp]: 1. `std::to_string` is C++11, `::to_string` is defined in `util/string.h`.
  2. It's `TOSTRING()`, but yes, I probably should.

File: src/network/connection.cpp [hunk 10]
```diff
@@ -2986,7 +2942,7 @@ float Connection::getLocalStat(rate_stat_type type)
 	return retval;
 }
 
-u16 Connection::createPeer(Address& sender, MTProtocols protocol, int fd)
+u16 Connection::createPeer(const Address &sender, size_t sock_id, MTProtocols protocol, int fd)
```
Comment (kwolekr) [inline src/network/connection.cpp]: Is sock_id really supposed to be a size_t?

**Comment by kwolekr** [comment]
All of these changes are obviously welcome, but kind of wish it didn't have to be so large.

File: src/ban.cpp (hunk 3 above)
Comment (ShadowNinja) [inline src/ban.cpp]: I find them about the same, but `std::ios::good` checks eofbit, so the eof check should be removed.

File: src/game.cpp (hunk 4 above)
Comment (ShadowNinja) [inline src/game.cpp]: Huh?  I'm shadowing something?

File: src/network/connection.cpp (hunk 8 above)
Comment (ShadowNinja) [inline src/network/connection.cpp]: It's modified.  Do you mean that you'd prefer a pointer parameter?

File: src/network/connection.cpp (hunk 9 above)
Comment (ShadowNinja) [inline src/network/connection.cpp]: I don't suppose it matters?

File: src/network/connection.cpp (hunk 10 above)
Comment (ShadowNinja) [inline src/network/connection.cpp]: Well, it should be a `std::vector<UDPSocket*>::size_type`, but I figured `size_t was close enough.

File: src/network/socket.cpp (hunk 7 above)
Comment (ShadowNinja) [inline src/network/socket.cpp]: I'm not sure if that's defined somewhere, but I'll remove the blocks (they're an artifact of a previous iteration of the code where they were necessary).

File: src/game.cpp (hunk 4 above)
Comment (kwolekr) [inline src/game.cpp]: listen()

File: src/network/connection.cpp (hunk 8 above)
Comment (kwolekr) [inline src/network/connection.cpp]: Yes.  If it's being modified it should be passed as a pointer.
If it's more advantageous syntax-wise to treat it as a reference, there's nothing stopping you from doing T &fooref = *fooarg;

File: src/network/connection.cpp (hunk 9 above)
Comment (kwolekr) [inline src/network/connection.cpp]: Functionality-wise, no.  But everywhere vectors are used (aside from the nerzhul list -> vector changes) they are treated as arrays.

File: src/network/connection.cpp (hunk 10 above)
Comment (kwolekr) [inline src/network/connection.cpp]: Okay then.  That's good enough for me.

**Comment by ShadowNinja** [comment]
@nerzhul, your build broke again.

File: src/network/connection.cpp (hunk 9 above)
Comment (ShadowNinja) [inline src/network/connection.cpp]: Well, I guess it's less likely that you'll accidentally decrease performance more than necessary if you change the socket container type to a list this way.  It's probably slightly less performant with a vector though.  Either way's fine with me.

**Comment by nerzhul** [comment]
Looking at this today, i will update Jenkins plugins

File: src/defaultsettings.cpp (hunk 6 above)
Comment (ShadowNinja) [inline src/defaultsettings.cpp]: Actually `TOSTRING()` won't work, because `*DEFAULT_FONT_SIZE` are defined with parenthesis around their values and that's included in the preprocessor-generated string.

The definitions could probably be changed though, I can't think of any unexpected behavior that would result from that.

Comment (Zeno-) [inline src/defaultsettings.cpp]: What's wrong with the original method (using stringstream)?

Comment (ShadowNinja) [inline src/defaultsettings.cpp]: @Zeno-: Just that it's longer.  A `stringstream` is still used internally by `to_string`.

**Comment by est31** [comment]
@nerzhul you are network maintainer, can you review this?

File: src/guiMainMenu.h [hunk 11]
```diff
@@ -25,37 +25,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 	std::string errormessage;
+
 	MainMenuData():
 		enable_public(false),
 		selected_world(0),
-		simple_singleplayer_mode(false),
-		errormessage("")
+		simple_singleplayer_mode(false)
 	{}
 };
```
Comment (nerzhul) [inline src/guiMainMenu.h]: Please do all those cleanups in a separate commit into master and rebase

**Comment by nerzhul** [comment]
Also, for the socket.{cpp,h} you have my approval for moving them in network directory and change the makefiles. This also permit to follow socket.{cpp,h} modifications better.

**Comment by ShadowNinja** [comment]
Rebased.

File: minetest.conf.example [hunk 12]
```diff
@@ -25,6 +25,11 @@
 #    Name of player, on a server this is the main admin
 #name =
 
+#    Network address to bind to when making outgoing connections.
+#    If you're running a server and have anouncing enabled the
```
Comment (est31) [inline minetest.conf.example]: announcing*

File: src/game.cpp (hunk 4 above)
Comment (est31) [inline src/game.cpp]: And `listen_addr` would make the meaning more clear.

File: minetest.conf.example (hunk 12 above)
Comment (ShadowNinja) [inline minetest.conf.example]: Actually it should probably be "announcement".

Comment (est31) [inline minetest.conf.example]: agree, thats better.

Comment (Zeno-) [inline minetest.conf.example]: Actually it should be "announce enabled", or "server_announce enabled"

**Comment by kwolekr** [comment]
Look, I support the idea behind this pull request, but seriously, enough with the 500000 line, 4000 file changes for one single feature.  Get rid of all the refactoring bits of this PR, and break it down into several discrete commits if necessary.  I am not convinced that adding multiple listen address support requires changes across 38 files.
:-1:  from me for now.  This is way too unwieldy.  Sorry.

**Comment by ShadowNinja** [comment]
@kwolekr: I've looked over it and I found four files that could be seperated (which I'll do), but most of it's needed.  Explanation:
Android.mk, CMakeLists.txt, network/CMakeLists.txt, httpfetch.cpp (socket.h move)
network/connection.cpp/h (Address API change, send packets to players from the same socket they connected to, other things to handle multiple sockets)
network/serverpackethandler.cpp, script/lua_api/l_server.cpp (Address API change)
Main menu, game.cpp/h, clientlauncher.cpp, gameparams.h, guiMainMenu.h, main.cpp, script/cpp_api/s_mainmenu.cpp, script/lua_api/l_mainmenu.cpp/h (Add listen option)
Man page (--listen)
minetest.conf.example, defaultsettings.cpp (listen setting)
client.cpp/h (remove global IPv6 bool)
server.cpp/h (^ and listen-related stuff)
socket.cpp/h, network/socket.cpp/h (Move and lots of refactoring)
unittest/test.h, unittest/test_connection.cpp, unittest/test_socket.cpp (Updating and unifying tests)

And related, but doesn't have to be in the same commit:
ban.cpp (normalization)

File: src/ban.cpp [hunk 13]
```diff
@@ -120,24 +130,23 @@ std::string BanManager::getBanName(const std::string &ip)
...
 {
 	MutexAutoLock lock(m_mutex);
-	m_ips[ip] = name;
+	m_ips[normalizeIp(ip)] = name;
 	m_modified = true;
 }
 
 void BanManager::remove(const std::string &ip_or_name)
 {
+	std::string norm = normalizeIp(ip_or_name);
```
Comment (est31) [inline src/ban.cpp]: it also "normalizes" the name. So if a good player's name is "NyanKitty", and the bad player "nyankitty" gets banned, and the admin unbans NyanKitty, does it then also unban nyankitty?

**Comment by est31** [comment]
Additional things, albeit smaller:
* `builtin/init.lua` : Variable rename, not at all related to the PR's goals
* `src/defaultsettings.cpp` : Nice that you do a reorder (my long term goal would be to split settings up to client / world / global or sth else), but can it be done in a separate commit?
* `src/unittest/test.h` : This can be separated as well.

**Comment by ShadowNinja** [comment]
@est31:
  * Hang on a sec, I'm doing that right now :-P
  * That too.
  * No, The parameter list has to be included so that you can tell the difference between the IPv4 test and the IPv6 test (they use the same method now since their code is almost identical).

**Comment by est31** [comment]
The serverlist.{cpp,h} changes can be exported as well.

**Comment by ShadowNinja** [comment]
@est31: And that too...

**Comment by est31** [comment]
What about doing the test.h change in a commit before the main commit? Then you can tell the difference with the main commit checked out.

File: src/ban.cpp (hunk 13 above)
Comment (ShadowNinja) [inline src/ban.cpp]: Yes, I've fixed this now.

**Comment by ShadowNinja** [comment]
I don't understand your second sentence.  And yes, it could be seperated, but it seems so trivial (it's only one line) that it doesn't really deserve its own commit.

File: src/guiEngine.cpp [hunk 14]
```diff
@@ -230,7 +230,7 @@ GUIEngine::GUIEngine(	irr::IrrlichtDevice* dev,
 /******************************************************************************/
 bool GUIEngine::loadMainMenuScript()
 {
-	// Try custom menu script (main_menu_path)
+	// Set main_menu_path
```
Comment (est31) [inline src/guiEngine.cpp]: How does this "set" the main_menu_path? It only reads it. I agree though, the old comment is misleading, so it should be changed.

**Comment by est31** [comment]
Note that when applying the patch, I get 4 trailing whitespaces, and 2 new blank lines at EOF. Perhaps do git rebase with `--whitespace=fix`.

**Comment by est31** [comment]
`src/script/lua_api/l_server.cpp` Has lots of whitespace only changes as well.

**Comment by est31** [comment]
https://github.com/minetest/minetest/pull/2604/files#diff-18513665750ef5adf42b5ec29e14162eR1453

why no reference?

**Comment by ShadowNinja** [comment]
@est31:
 * Huh, I'll try to find those whitespace errors, although grep isn't being cooperative...
 * And lots of functionality changes.  Spliting it up will be hard and make changing that file potentially cause conflicts.
  * Because that's how it was before.  Changed.

**Comment by est31** [comment]
Why would it cause conflicts to only split out the whitespace only style fixes/adjustments?

**Comment by ShadowNinja** [comment]
@est31: Because if I modify one of the commits, and the other one follows it, I could change its context lines.

**Comment by est31** [comment]
http://stackoverflow.com/a/2118107

**Comment by proller** [comment]
why multiple listen addresses needed?
when not enough listen :: or 0.0.0.0 ?

**Comment by ShadowNinja** [comment]
@proller: Because then you have to support only one.  On O.S.es where IPV6_V6ONLY defaults to false (some Linux distros) you could do both by selecting the IPv6 address, then but the ban list would add entries with `::ffff:1.2.3.4` which would cease to work if you used a regular IPv4 socket.

**Comment by ShadowNinja** [comment]
Rebased.

File: src/network/socket.cpp [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+		addr->v6.sin6_addr = host.v6;
+		return sizeof(addr->v6);
+	}
+	return 0;
+}
+
+bool Address::isLoopback() const
+{
+	switch (family) {
+	case AF_INET: return ntohl(host.v4.s_addr) == INADDR_LOOPBACK;
```
Comment (numberZero) [inline src/network/socket.cpp]: AFAIK the whole `127.0.0.0/8` network is defined as loopback. `INADDR_LOOPBACK` is just a single address in it (`127.0.0.1`).
Note: `127.0.0.2` works perfectly on Linux. Not sure about Windows, it may have some problems with that.

**Comment by paramat** [comment]
May be best to leave the refactor / code style stuff to another PR to make this easier? should at least be a separate commit.

**Comment by paramat** [comment]
@ShadowNinja please can you attend to this or close it?
Please split off anything that can be into new PRs so we can get those merged.

**Comment by paramat** [comment]
@ShadowNinja assuming adoption needed.

**Comment by paramat** [comment]
Anyone interested in adopting this? If so is any of this useful seeing how old it is and the conflicts?
Closing in 1 month unless there is a reason not to.

File: src/script/lua_api/l_server.cpp [hunk 16]
```diff
@@ -174,90 +171,71 @@ int ModApiServer::l_get_player_information(lua_State *L)
...
+	int f = addr.getFamily();
+	lua_pushnumber(L, f == AF_INET ? 4 :
+			f == AF_INET6 ? 6 :
+			0);
+	lua_setfield(L, table, "ip_version");
 
-	lua_pushstring(L,"min_rtt");
 	lua_pushnumber(L, min_rtt);
-	lua_settable(L, table);
+	lua_setfield(L, table, "min_rtt");
```
Comment (HybridDog) [inline src/script/lua_api/l_server.cpp]: ‭

**Comment by paramat** [comment]
IRC http://irc.minetest.net/minetest-dev/2017-10-07#i_5101918

18:24 paramat      what about #2604 ? should we close due to amount of conflicts?
18:25 paramat      i've marked all your open PRs for adoption
18:26 ShadowNinja  paramat: I'd rather not.  I think it's still usefull, and it should be merged at some point.
18:27 paramat      it would have to be reworked, it's probably easier to start again from nothing due to conflicts
18:27 paramat      it's only open to be seen for adoption

**Comment by rubenwardy** [comment]
This is very old with massive rebase problems, I also see this as a low-priority feature on the scale of things. If you'd like this then please reopen once you've updated and fixed any issues

**Comment by paramat** [comment]
^ I support the close.
